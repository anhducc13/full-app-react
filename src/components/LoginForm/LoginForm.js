import React, { useState } from 'react';
import { setGlobal } from 'reactn';
import Avatar from '@material-ui/core/Avatar';
import CssBaseline from '@material-ui/core/CssBaseline';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { Link as RouteLink } from 'react-router-dom';
import { Button as ButtonCustom } from '../shared/Button';
import { InputText } from '../shared/InputText';
import { authService } from 'services';
import { validateUsername, validatePassword } from 'helpers/validators';
import { errorSwal } from 'helpers/swal';
import { toast } from 'react-toastify';

const useStyles = makeStyles(theme => ({
  '@global': {
    body: {
      backgroundColor: theme.palette.common.white,
    },
  },
  paper: {
    marginTop: theme.spacing(2),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: 60,
    height: 60,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}));


export const LoginForm = (props) => {
  const classes = useStyles();

  const initTextField = {
    value: '',
    error: '',
    valid: false,
  };

  const [usernameField, setUsernameField] = useState(initTextField);
  const [passwordField, setPasswordField] = useState(initTextField);

  const onChangeUsername = (e) => {
    let { value, error, valid } = usernameField;
    value = e.target.value;
    const outputValidate = validateUsername({ value, error, valid });
    setUsernameField(outputValidate);
  }

  const onChangePassword = (e) => {
    let { value, error, valid } = passwordField;
    value = e.target.value;
    const outputValidate = validatePassword({ value, error, valid });
    setPasswordField(outputValidate);
  }

  const resetForm = () => {
    setUsernameField(initTextField);
    setPasswordField(initTextField);
  }

  const onSubmitForm = (e) => {
    e.preventDefault();
    const params = {
      "username": usernameField.value,
      "password": passwordField.value
    }
    setGlobal({
      loading: true
    });
    authService.login(params)
      .then(res => {
        console.log(res)
        setGlobal({
          loading: false
        });
        if (res.status === 200) {
          const userInfoLogin = res.data;
          localStorage.setItem('userInfoLogin', JSON.stringify(userInfoLogin));
          toast.success(`Xin chào ${userInfoLogin['username']} quay trở lại`);
          props.history.push('/');
        }
      })
      .catch(err => {
        setGlobal({
          loading: false
        });
        if (err.response && err.response.data)
          errorSwal({
            title: 'Có lỗi xảy ra!',
            content: err.response.data.message || ''
          }, resetForm)
        else
          props.history.push('/error');
      })
  }


  return (
    <Container component="main" maxWidth="sm">
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          Đăng nhập
        </Typography>
        <form className={classes.form} onSubmit={onSubmitForm} noValidate>
          <InputText
            textField={usernameField}
            name="username"
            label="Tên đăng nhập"
            type="text"
            onChange={onChangeUsername}
          />
          <InputText
            textField={passwordField}
            name="password"
            label="Mật khẩu"
            type="password"
            onChange={onChangePassword}
          />
          <ButtonCustom
            displayText="Đăng nhập"
            disabled={!usernameField.valid || !passwordField.valid}
          />
          <Grid container>
            <Grid item xs>
              <Link to="/quen-mat-khau" variant="body2" component={RouteLink}>
                Quên mật khẩu?
              </Link>
            </Grid>
            <Grid item>
              <Link to="/dang-ky" variant="body2" component={RouteLink}>
                Chưa có tài khoản? Đăng ký
              </Link>
            </Grid>
          </Grid>
        </form>
      </div>
    </Container>
  );
}