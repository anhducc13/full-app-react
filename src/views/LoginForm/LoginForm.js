import React from 'react';
import Avatar from '@material-ui/core/Avatar';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import TextField from '@material-ui/core/TextField';
import Link from '@material-ui/core/Link';
import Grid from '@material-ui/core/Grid';
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import { Link as RouteLink} from 'react-router-dom';


const useStyles = makeStyles(theme => ({
  '@global': {
    body: {
      backgroundColor: theme.palette.common.white,
    },
  },
  paper: {
    marginTop: theme.spacing(2),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
    width: 60,
    height: 60,
  },
  form: {
    width: '100%', // Fix IE 11 issue.
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}));

export const LoginForm = (props) => {
  const classes = useStyles();
  const { usernameField, passwordField } = props.textField;
  const { onChangeUsername, onChangePassword } = props.handleChange
  const submitForm = props.handleSubmit

  return (
    <Container component="main" maxWidth="sm">
      <CssBaseline />
      <div className={classes.paper}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          Login
        </Typography>
        <form className={classes.form} onSubmit={submitForm} noValidate>
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            id="username"
            label="Username"
            name="username"
            autoFocus
            value={usernameField.value}
            error={usernameField.error !== ''}
            onChange={onChangeUsername}
            helperText={usernameField.valid ? "" : usernameField.error}
          />
          <TextField
            variant="outlined"
            margin="normal"
            required
            fullWidth
            name="password"
            label="Password"
            type="password"
            id="password"
            value={passwordField.value}
            error={passwordField.error !== ''}
            onChange={onChangePassword}
            helperText={passwordField.error ? "" : passwordField.error}
          />
          {/* <FormControlLabel
              control={<Checkbox value="remember" color="primary" />}
              label="Remember me"
            /> */}
          <Button
            size="large"
            type="submit"
            fullWidth
            variant="contained"
            color="primary"
            className={classes.submit}
            disabled={!usernameField.valid || !passwordField.valid}
          >
            Sign In
          </Button>
          <Grid container>
            <Grid item xs>
              <Link to="/forgot-password" variant="body2" component={RouteLink}>
                Forgot password?
              </Link>
            </Grid>
            <Grid item>
              <Link to="/register" variant="body2" component={RouteLink}>
                {"Don't have an account? Register"}
              </Link>
            </Grid>
          </Grid>
        </form>
      </div>
    </Container>
  );
}