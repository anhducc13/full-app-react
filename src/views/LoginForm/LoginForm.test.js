import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { LoginForm } from './LoginForm';


const initTextField = {
  value: '',
  error: '',
  valid: false,
}
const defaultProps = {
  textField: {
    usernameField: initTextField,
    passwordField: initTextField,
  },
  handleChange: {
    onChangeUsername: jest.fn(),
    onChangePassword: jest.fn(),
  },
  handleSubmit: jest.fn()
}

describe('LoginForm', () => {
  it('should render correctly', () => {
    const component = shallow(
        <LoginForm {...defaultProps}/>
    );
    expect(component).toMatchSnapshot();
    component.unmount();
  });
});

describe('Login form tests', () => {
  
  it('have username input & password input', () => {
    const LoginFormComponent = mount(
      <MemoryRouter>
        <LoginForm {...defaultProps} />
      </MemoryRouter>
    );
    const usernameInput = LoginFormComponent.find('input[name="username"]');
    expect(usernameInput).toBeDefined();
    expect(usernameInput).toHaveLength(1);
    const passwordInput = LoginFormComponent.find('input[name="password"]');
    expect(passwordInput).toBeDefined();
    expect(passwordInput).toHaveLength(1);
    LoginFormComponent.unmount();
  });

  it('have a button submit and default is disabled', () => {
    const LoginFormComponent = mount(
      <MemoryRouter>
        <LoginForm {...defaultProps} />
      </MemoryRouter>
    );
    const buttonSubmit = LoginFormComponent.find('button[type="submit"][disabled=true]');
    expect(buttonSubmit).toBeDefined();
    expect(buttonSubmit).toHaveLength(1);
    LoginFormComponent.unmount();
  });

  it('have 2 link, 1 to register, 1 to forgot password', () => {
    const LoginFormComponent = mount(
      <MemoryRouter>
        <LoginForm {...defaultProps} />
      </MemoryRouter>
    );
    const linkToRegister = LoginFormComponent.find('a[href="/register"]');
    expect(linkToRegister).toBeDefined();
    expect(linkToRegister).toHaveLength(1);

    const linkToForgotPassword = LoginFormComponent.find('a[href="/forgot-password"]');
    expect(linkToForgotPassword).toBeDefined();
    expect(linkToForgotPassword).toHaveLength(1);
    LoginFormComponent.unmount();
  });

  it('onChangeUsername have been call when change input username', () => {
    const wrapper = mount(
      <MemoryRouter>
        <LoginForm {...defaultProps} />
      </MemoryRouter>
    );
    wrapper.find('input[name="username"]').simulate('change');
    expect(defaultProps.handleChange.onChangeUsername).toHaveBeenCalled();
    wrapper.unmount();
  });

  it('onChangePassword have been call when change input password', () => {
    const wrapper = mount(
      <MemoryRouter>
        <LoginForm {...defaultProps} />
      </MemoryRouter>
    );
    wrapper.find('input[name="password"]').simulate('change');
    expect(defaultProps.handleChange.onChangePassword).toHaveBeenCalled();
    wrapper.unmount();
  });

});