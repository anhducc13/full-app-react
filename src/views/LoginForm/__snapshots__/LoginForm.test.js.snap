// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should render correctly 1`] = `
<WithStyles(ForwardRef(Container))
  component="main"
  maxWidth="sm"
>
  <CssBaseline />
  <div
    className="makeStyles-paper-1"
  >
    <WithStyles(ForwardRef(Avatar))
      className="makeStyles-avatar-2"
    >
      <LockOutlinedIcon />
    </WithStyles(ForwardRef(Avatar))>
    <WithStyles(ForwardRef(Typography))
      component="h1"
      variant="h4"
    >
      Login
    </WithStyles(ForwardRef(Typography))>
    <form
      className="makeStyles-form-3"
      noValidate={true}
      onSubmit={[MockFunction]}
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        error={false}
        fullWidth={true}
        helperText=""
        id="username"
        label="Username"
        margin="normal"
        name="username"
        onChange={[MockFunction]}
        required={true}
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(TextField))
        error={false}
        fullWidth={true}
        helperText=""
        id="password"
        label="Password"
        margin="normal"
        name="password"
        onChange={[MockFunction]}
        required={true}
        type="password"
        value=""
        variant="outlined"
      />
      <WithStyles(ForwardRef(Button))
        className="makeStyles-submit-4"
        color="primary"
        disabled={true}
        fullWidth={true}
        size="large"
        type="submit"
        variant="contained"
      >
        Sign In
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Grid))
        container={true}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          xs={true}
        >
          <WithStyles(ForwardRef(Link))
            component={[Function]}
            to="/forgot-password"
            variant="body2"
          >
            Forgot password?
          </WithStyles(ForwardRef(Link))>
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <WithStyles(ForwardRef(Link))
            component={[Function]}
            to="/register"
            variant="body2"
          >
            Don't have an account? Register
          </WithStyles(ForwardRef(Link))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </form>
  </div>
</WithStyles(ForwardRef(Container))>
`;
